File: qt/gen_qtoolbutton.go
package qt

import (
	"unsafe"
)

type QToolButton__ToolButtonPopupMode int

const (
	QToolButton__DelayedPopup    QToolButton__ToolButtonPopupMode = 0
	QToolButton__MenuButtonPopup QToolButton__ToolButtonPopupMode = 1
	QToolButton__InstantPopup    QToolButton__ToolButtonPopupMode = 2
)

type QToolButton struct {
	h          uintptr
	isSubclass bool
}

// NewQToolButton constructs a new QToolButton object.
func NewQToolButton(parent *QWidget) *QToolButton {
	ret := newQToolButton(QToolButton_new(parent.cPointer()))
	ret.isSubclass = true
	return ret
}

// NewQToolButton2 constructs a new QToolButton object.
func NewQToolButton2() *QToolButton {
	ret := newQToolButton(QToolButton_new2())
	ret.isSubclass = true
	return ret
}

func (this *QToolButton) MetaObject() *QMetaObject {
	return newQMetaObject(QToolButton_MetaObject(this.h))
}

func (this *QToolButton) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QToolButton_Metacast(this.h, param1_Cstring))
}

func QToolButton_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QToolButton_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QToolButton) SizeHint() *QSize {
	_goptr := newQSize(QToolButton_SizeHint(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QToolButton) MinimumSizeHint() *QSize {
	_goptr := newQSize(QToolButton_MinimumSizeHint(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QToolButton) ToolButtonStyle() ToolButtonStyle {
	return (ToolButtonStyle)(QToolButton_ToolButtonStyle(this.h))
}

func (this *QToolButton) ArrowType() ArrowType {
	return (ArrowType)(QToolButton_ArrowType(this.h))
}

func (this *QToolButton) SetArrowType(typeVal ArrowType) {
	QToolButton_SetArrowType(this.h, (int)(typeVal))
}

func (this *QToolButton) SetMenu(menu *QMenu) {
	QToolButton_SetMenu(this.h, menu.cPointer())
}

func (this *QToolButton) Menu() *QMenu {
	return newQMenu(QToolButton_Menu(this.h))
}

func (this *QToolButton) SetPopupMode(mode ToolButtonPopupMode) {
	QToolButton_SetPopupMode(this.h, mode)
}

func (this *QToolButton) PopupMode() ToolButtonPopupMode {
	xxxxxxxxx
}

func (this *QToolButton) DefaultAction() *QAction {
	return newQAction(QToolButton_DefaultAction(this.h))
}

func (this *QToolButton) SetAutoRaise(enable bool) {
	QToolButton_SetAutoRaise(this.h, (bool)(enable))
}

func (this *QToolButton) AutoRaise() bool {
	return (bool)(QToolButton_AutoRaise(this.h))
}

func (this *QToolButton) ShowMenu() {
	QToolButton_ShowMenu(this.h)
}

func (this *QToolButton) SetToolButtonStyle(style ToolButtonStyle) {
	QToolButton_SetToolButtonStyle(this.h, (int)(style))
}

func (this *QToolButton) SetDefaultAction(defaultAction *QAction) {
	QToolButton_SetDefaultAction(this.h, defaultAction.cPointer())
}

func (this *QToolButton) Triggered(param1 *QAction) {
	QToolButton_Triggered(this.h, param1.cPointer())
}

func (this *QToolButton) OnTriggered(slot func(param1 *QAction)) {
	QToolButton_connect_Triggered(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QToolButton_Triggered
func miqt_exec_callback_QToolButton_Triggered(cb intptr_t, param1 *QAction) {
	gofunc, ok := cgo.Handle(cb).Value().(func(param1 *QAction))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := newQAction(param1)

	gofunc(slotval1)
}

func QToolButton_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QToolButton_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QToolButton_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QToolButton_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QToolButton) callVirtualBase_MetaObject() *QMetaObject {
	return newQMetaObject(QToolButton_virtualbase_MetaObject(unsafe.Pointer(this.h)))
}

func (this *QToolButton) OnMetaObject(slot func(super func() *QMetaObject) *QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QToolButton_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QToolButton_MetaObject
func miqt_exec_callback_QToolButton_MetaObject(self QToolButton, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *QMetaObject) *QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QToolButton{h: self}).callVirtualBase_MetaObject)

	return virtualReturn.cPointer()
}

func (this *QToolButton) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QToolButton_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QToolButton) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QToolButton_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QToolButton_Metacast
func miqt_exec_callback_QToolButton_Metacast(self QToolButton, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)

	virtualReturn := gofunc((&QToolButton{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}

File: qt/gen_qtoolbutton.cpp
// +build ignore

#include <QAbstractButton>
#include <QAction>
#include <QActionEvent>
#include <QEnterEvent>
#include <QEvent>
#include <QMenu>
#include <QMetaObject>
#include <QMouseEvent>
#include <QObject>
#include <QPaintDevice>
#include <QPaintEvent>
#include <QPoint>
#include <QSize>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QStyleOptionToolButton>
#include <QTimerEvent>
#include <QToolButton>
#include <QWidget>
#include <qtoolbutton.h>
#include "gen_qtoolbutton.h"
class MiqtVirtualQToolButton : public virtual QToolButton {
public:
MiqtVirtualQToolButton(QWidget* parent): QToolButton(parent) {};
MiqtVirtualQToolButton(): QToolButton() {};

virtual ~MiqtVirtualQToolButton() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QToolButton::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QToolButton_MetaObject(const_cast<MiqtVirtualQToolButton*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QToolButton::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QToolButton::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QToolButton_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QToolButton::qt_metacast(param1);

	}
};
QToolButton* QToolButton_new(QWidget* parent) {
return new MiqtVirtualQToolButton(parent);
}
QToolButton* QToolButton_new2() {
return new MiqtVirtualQToolButton();
}
void QToolButton_virtbase(QToolButton* src
, QAbstractButton** outptr_QAbstractButton
) {
*outptr_QAbstractButton = static_cast<QAbstractButton*>(src);
}
QMetaObject* QToolButton_MetaObject(const QToolButton* self) {
	return (QMetaObject*) self->metaObject();
}
void* QToolButton_Metacast(QToolButton* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QToolButton_Tr(const char* s) {
	QString _ret = QToolButton::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
QSize* QToolButton_SizeHint(const QToolButton* self) {
	return new QSize(self->sizeHint());
}
QSize* QToolButton_MinimumSizeHint(const QToolButton* self) {
	return new QSize(self->minimumSizeHint());
}
int QToolButton_ToolButtonStyle(const QToolButton* self) {
	Qt::ToolButtonStyle _ret = self->toolButtonStyle();
	return static_cast<int>(_ret);
}
int QToolButton_ArrowType(const QToolButton* self) {
	Qt::ArrowType _ret = self->arrowType();
	return static_cast<int>(_ret);
}
void QToolButton_SetArrowType(QToolButton* self, int typeVal) {
	self->setArrowType(static_cast<Qt::ArrowType>(typeVal));
}
void QToolButton_SetMenu(QToolButton* self, QMenu* menu) {
	self->setMenu(menu);
}
QMenu* QToolButton_Menu(const QToolButton* self) {
	return self->menu();
}
void QToolButton_SetPopupMode(QToolButton* self, ToolButtonPopupMode mode) {
	self->setPopupMode(mode);
}
ToolButtonPopupMode QToolButton_PopupMode(const QToolButton* self) {
	return self->popupMode();
}
QAction* QToolButton_DefaultAction(const QToolButton* self) {
	return self->defaultAction();
}
void QToolButton_SetAutoRaise(QToolButton* self, bool enable) {
	self->setAutoRaise(enable);
}
bool QToolButton_AutoRaise(const QToolButton* self) {
	return self->autoRaise();
}
void QToolButton_ShowMenu(QToolButton* self) {
	self->showMenu();
}
void QToolButton_SetToolButtonStyle(QToolButton* self, int style) {
	self->setToolButtonStyle(static_cast<Qt::ToolButtonStyle>(style));
}
void QToolButton_SetDefaultAction(QToolButton* self, QAction* defaultAction) {
	self->setDefaultAction(defaultAction);
}
void QToolButton_Triggered(QToolButton* self, QAction* param1) {
	self->triggered(param1);
}
void QToolButton_connect_Triggered(QToolButton* self, intptr_t slot) {
	MiqtVirtualQToolButton::connect(self, static_cast<void (QToolButton::*)(QAction*)>(&QToolButton::triggered), self, [=](QAction* param1) {
		QAction* sigval1 = param1;
		miqt_exec_callback_QToolButton_Triggered(slot, sigval1);
	});
}
struct miqt_string QToolButton_Tr2(const char* s, const char* c) {
	QString _ret = QToolButton::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QToolButton_Tr3(const char* s, const char* c, int n) {
	QString _ret = QToolButton::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QToolButton_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQToolButton*>( (QToolButton*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QToolButton_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQToolButton*)(self) )->virtualbase_MetaObject();
}
void QToolButton_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQToolButton*>( (QToolButton*)(self) )->handle__Metacast = slot;
}
void* QToolButton_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQToolButton*)(self) )->virtualbase_Metacast(param1);
}
void QToolButton_Delete(QToolButton* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQToolButton*>( self );
	} else {
		delete self;
	}
}
